src/machine.ts

```diff
@@ -0,0 +1,66 @@
+import { Path, PathType } from './path';
+
+export class State {
+  x: number;
+  y: number;
+  z: number;
+  e: number;
+  tool: number;
+  units: 'mm' | 'in';
+
+  static get initial(): State {
+    const state = new State();
+    Object.assign(state, { x: 0, y: 0, z: 0, e: 0, tool: 0, units: 'mm' });
+    return state;
+  }
+}
+
+export class Machine {
+  paths: Path[];
+  state: State;
+
+  constructor(state?: State) {
+    this.paths = [];
+    this.state = state || State.initial;
+  }
+
+  isPlanar(): boolean {
+    return (
+      this.paths.find(
+        (path) =>
+          path.travelType === PathType.Extrusion && path.vertices.some((_, i, arr) => i % 3 === 2 && arr[i] !== arr[2])
+      ) === undefined
+    );
+  }
+
+  layers(): Path[][] | null {
+    if (!this.isPlanar()) {
+      return null;
+    }
+
+    const layers: Path[][] = [];
+    let currentLayer: Path[] = [];
+
+    this.paths.forEach((path) => {
+      if (path.travelType === PathType.Extrusion) {
+        currentLayer.push(path);
+      } else {
+        if (path.vertices.some((_, i, arr) => i % 3 === 2 && arr[i] !== arr[2])) {
+          layers.push(currentLayer);
+          currentLayer = [];
+        }
+        currentLayer.push(path);
+      }
+    });
+
+    return layers;
+  }
+
+  extrusions(): Path[] {
```

Comment:
this looks very expensive. It will filter all paths every time and allocate a subset.

Reply:
It is, but it's only ran once in the flow. This is a first lazy implementation before I consider adding an index approach that would work better for different rendering modes.